### PR TITLE
feat(storage): block format epoch dictionary encoding

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by wcy-fdu** [comment]
Any benchmark result?

File: src/storage/src/hummock/sstable/block.rs [hunk 1]
```diff
@@ -149,6 +166,8 @@ pub struct Block {
 
     /// Restart points.
     restart_points: Vec<RestartPoint>,
+
+    epoch_dictionary: EpochDictionary,
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:170]: need more documentation on what the structure of this dictionary is and how we use it.

File: src/storage/src/hummock/sstable/block.rs [hunk 2]
```diff
@@ -445,17 +512,29 @@ impl BlockBuilder {
         #[cfg(debug_assertions)]
         self.debug_valid();
 
+        let epoch = full_key.epoch;
+        self.epoch_dictionary.insert(epoch);
+
         let mut key: BytesMut = Default::default();
         full_key.encode_into_without_table_id(&mut key);
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:519]: We don't need to encode epoch into `key`. We can avoid slicing in L534 and L567

File: src/storage/src/hummock/sstable/block.rs [hunk 3]
```diff
@@ -473,6 +552,7 @@ impl BlockBuilder {
             let offset = self.buf.len() as u32;
 
             self.restart_points.push(offset);
+            self.epoch_group.push(Vec::default());
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:555]: We can use `Vec::with_capacity` since we know the number of epochs per group beforehand.

File: src/storage/src/hummock/sstable/block.rs [hunk 4]
```diff
@@ -697,7 +817,7 @@ mod tests {
         builder.add(construct_full_key_struct(0, b"k3", 3), b"v03");
         builder.add(construct_full_key_struct(0, b"k4", 4), b"v04");
         let capacity = builder.uncompressed_block_size();
-        assert_eq!(capacity, builder.approximate_len() - 9);
+        // assert_eq!(capacity, builder.approximate_len() - 9);
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:820]: need to fix

File: src/storage/src/hummock/sstable/block.rs [hunk 5]
```diff
@@ -569,7 +660,40 @@ impl BlockBuilder {
         self.buf
             .put_u32_le(self.restart_points_type_index.len() as u32);
 
+        let mut epoch_index: usize = 0;
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:663]: Please update the doc of `build` in L629 accordingly

File: src/storage/src/hummock/sstable/block.rs [hunk 6]
```diff
@@ -569,7 +660,40 @@ impl BlockBuilder {
...
+            .map(|epoch| {
+                epoch_index += 1;
+                (epoch, epoch_index)
+            })
+            .collect::<BTreeMap<HummockEpoch, usize>>();
+
+        // encode epoch_group
+        for group in &self.epoch_group {
+            let group_len = group.len();
+            self.buf.put_u16_le(group_len as u16);
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:677]: Although highly unlikely, the possibility of group_len overflow still exists. How about focing a block swtich when the number of keys added to the block >= u16::MAX?

File: src/storage/src/hummock/sstable/block.rs [hunk 7]
```diff
@@ -569,7 +660,40 @@ impl BlockBuilder {
...
+            for epoch in group {
+                let epoch_index = epoch_dictionary.get(epoch).unwrap();
+                self.buf.put_u16_le(*epoch_index as u16);
+            }
+        }
+
+        // epoch_group == resatrt_points count , not need to record
+        self.buf.put_u32_le(self.entry_count as u32);
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:685]: Based on the comment above, entry_count can be u16?

File: src/storage/src/hummock/sstable/block.rs [hunk 8]
```diff
@@ -569,7 +660,40 @@ impl BlockBuilder {
...
+        self.buf.put_u16_le(self.epoch_group.len() as u16);
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/block.rs:686]: Based on the comment in L684, why do we need to encode this?

File: src/storage/src/hummock/sstable/block.rs (hunk 3 above)
Comment (Li0k) [reply src/storage/src/hummock/sstable/block.rs:575]: we cant not know the number of key per restart_point due to the Flexible length encoding, i will add some document

**Comment by Li0k** [comment]
TLDR:
1. Introduces complexity and some performance degradation, with insignificant space savings
2. Version compatibility issues

@hzxa21 @wcy-fdu @Little-Wallace I suggest we can drop this feat, what's your opinion?

I executed two scenarios for testing
#### test 1
- key num 5w per block
- k-v pair (200B, 400B)
- restart_point_interval = 16

main cost
image

branch cost
image

#### test 2
- key num 1000 per block
- k-v pair 100B
- restart_point_interval = 16

main cost
image

branch cost 
image

**Comment by wcy-fdu** [comment]
From result of test 2:
- block size reduced to 92% of previous
- latency increased by ten percent

**Comment by wcy-fdu** [comment]
So you mean the impact of latency is greater than the benefits of block compression?  It's a trade off🤔

**Comment by Li0k** [comment]
> So you mean the impact of latency is greater than the benefits of block compression? It's a trade off🤔
> 
> Maybe we can hold this PR for a while, after #8584 is implemented, we can compare the benefits of different compression methods horizontally, and think about whether to let users choose whether to compress.

Yes，its trade-off between space with perform, just that I don't think the benefits are proportional
